packages/cli-e2e/utils/terminal/do.ts

```diff
@@ -0,0 +1,165 @@
+    new Promise<void>((resolve) => {
+      eventListener = this.addTerminalListener(
+        untilCondition,
+        (_process: ChildProcessWithoutNullStreams, resolve: () => void) => {
+          this.exitFunction!();
+          resolve();
+        },
+        resolve
+      );
+    }).then(() => {
+      this.removeListener(eventListener);
+    });
+    return this.promise;
+  }
+
+  /**
+   * Ensure the `callback` of the do is executed only once.
+   * An exit condition should always be specified, either with `until` or `once`.
+   * @returns {Promise} an awaitable promise that will resolve when the condition of the `Where` is satisfied and the callback of the `Do` is called.
+   */
+  public once(): Promise<void> {
+    /**
+     * Once hooks using exactly the same condition as the do. It ensures to be executed after the `doCallback` because
+     * Process & Stream both implements the EventEmitter class and uses it to broadcast on `on`.
+     * The EventEmitter class also ensure that the listener are called in FIFO order, which is key for this to work.
+     */
+    return this.until(this.action.condition);
+  }
+
+  /**
+   * What to do when the condition is fulfilled
+   * @param callback the function to execute.
+   *  It receives the process that fulfilled the condition
+   *  (or the process owning the stream that fulfilled the condition).
+   *  in parameter
+   * @returns a promise, extended with an 'until' function.
+   */
+  private do(callback: DoCallback = () => {}) {
+    return (resolve: () => void) => {
+      const eventListener = this.addTerminalListener(
+        this.action.condition,
+        async (process: ChildProcessWithoutNullStreams) => {
+          await callback(process);
+        },
+        resolve
+      );
+
+      this.exitFunction = () => {
+        this.removeListener(eventListener);
+        resolve();
+      };
+    };
+  }
+
+  private removeListener(eventListener: EventListenerObject) {
+    eventListener.target.off(eventListener.event, eventListener.listener);
+  }
+
+  private getTarget(): Readable | ChildProcessWithoutNullStreams {
+    return this.action.target === 'process'
+      ? this.action.process
+      : this.action.process[this.action.target];
+  }
+
+  private addTerminalListener(
+    condition: Condition,
+    callback: (
+      process: ChildProcessWithoutNullStreams,
+      resolve: () => void
+    ) => void,
+    resolve: () => void
+  ) {
+    const target: Readable | ChildProcessWithoutNullStreams = this.getTarget();
+
+    let onParams: [string, (data: Buffer) => void] | undefined;
+    if (isConditionRegExp(condition)) {
+      onParams = [
+        'data',
+        (data: Buffer) => {
+          if (stripAnsi(data.toString()).replace(/\n/g, '').match(condition)) {
+            callback(this.action.process, resolve);
+          }
+        },
+      ];
+    }
+
+    if (isConditionString(condition)) {
+      onParams = [condition, () => callback(this.action.process, resolve)];
```

Comment:
No because when a condition is a string, you listen to the event that is named as the condition, meaning that
```ts
new terminal(...spawnParams).when('potato').on('stdout').do(()=>console.log('foo!');
```

is the same as:
```ts
spawn(...spawnParams).stdout.on('potato',()=>console.log('foo!'));
```

and **not** the same as

```ts
spawn(...spawnParams)
  .stdout
  .on('data', () =>
     data.indexOf('potato')!==-1 ?
     console.log('foo!') :
     ()=>{}
  );
```